dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/McpRequestTracker.java

```diff
@@ -67,19 +68,19 @@ public Cancellation getOngoingRequestCancellation(ExecutionRequestId id) {
      * Will skip cancellation if the server is in stateless mode.
      * request is cancelled with a fixed reason: {@code "Session cancelled"}
      */
-    public void cancelSessionRequests(McpSession session) {
-
-        Boolean stateless = mcpConfigService.run(config -> config.isStateless()).orElse(false);
+    public void cancelSessionRequests(String sessionId) {
+        Boolean stateless = mcpConfigService.run(McpConfiguration::isStateless).orElse(false);
         if (Boolean.TRUE.equals(stateless)) {
             return;
         }
+        for (Map.Entry<ExecutionRequestId, CancellationImpl> entry : ongoingRequests.entrySet()) {
 
-        for (ExecutionRequestId id : session.getActiveRequests()) {
-            Cancellation cancellation = ongoingRequests.remove(id.toString());
-            if (cancellation instanceof CancellationImpl) {
-                ((CancellationImpl) cancellation).cancel(Optional.of("Session cancelled"));
+            if (sessionId.equals(entry.getKey().sessionId())) {
+                Cancellation cancellation = ongoingRequests.remove(entry.getKey());
+                if (cancellation instanceof CancellationImpl impl) {
+                    impl.cancel(Optional.of("Session cancelled"));
+                }
             }
         }
     }
```

Comment:
I really don't like this. We don't want to have to iterate over all in-flight requests to search for the ones which apply to a given session.